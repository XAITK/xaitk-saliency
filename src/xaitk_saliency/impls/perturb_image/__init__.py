"""Module for all implementations of the PerturbImage interface"""
