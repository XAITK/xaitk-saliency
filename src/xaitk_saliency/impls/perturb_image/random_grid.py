from typing import Any, Dict, Optional, Tuple

import numpy as np
from skimage.transform import resize
from smqtk_descriptors.utils import parallel_map

from xaitk_saliency import PerturbImage


class RandomGrid(PerturbImage):
    """
    Generate masks using a random grid of set cell size. If the chosen cell
    size does not divide an image evenly, then the grid is over-sized and
    the resulting mask is centered and cropped. Each mask is also shifted
    randomly by a maximum of half the cell size in both x and y.

    This method is based on RISE (http://bmvc2018.org/contents/papers/1064.pdf)
    but aims to address the changing cell size, given images of different
    sizes, aspect of that implementation. This method keeps cell size constant
    and instead adjusts the overall grid size for different sized images.

    :param n: Number of masks to generate.
    :param s: Dimensions of the grid cells in pixels. E.g. (3, 4) would use
        a grid of 3x4 pixel cells.
    :param p1: Probability of a grid cell being set to 1 (not occluded).
        This should be a float value in the [0, 1] range.
    :param seed: A seed to use for the random number generator, allowing
        for masks to be reproduced.
    :param threads: Number of threads to use when generating masks. If this
        is <=0 or None, no threading is used and processing is performed
        in-line serially.
    """

    def __init__(
        self,
        n: int,
        s: Tuple[int, int],
        p1: float,
        seed: Optional[int] = None,
        threads: Optional[int] = None,
    ):
        self.n = n
        self.s = s
        self.p1 = p1
        self.seed = seed
        self.threads = threads

    def perturb(
        self,
        ref_img: np.ndarray,
    ) -> np.ndarray:
        num_masks = self.n
        s = np.array(self.s)
        p1 = self.p1
        threads = self.threads
        grid_rng = np.random.default_rng(self.seed)
        shift_rng = np.random.default_rng(self.seed)

        img_shape = np.array(ref_img.shape[:2])

        # Pad grid by one cell in both dimensions to allow for shifting
        grid_size = (img_shape // s) + 1

        # Additional padding if cell size does not divide the image evenly
        # This is required to allow shifting at full range
        grid_size += img_shape % s != 0

        grid = grid_rng.random((num_masks, *grid_size)) < p1
        grid = grid.astype(np.float32)

        masks = np.empty((num_masks, *img_shape), dtype=grid.dtype)

        up_size = grid_size * s

        # Offsets to center mask on grid
        center_y, center_x = (up_size - img_shape) // 2

        # Maximum shifting offsets
        max_shift_y, max_shift_x = s // 2

        # Expand index for efficiency
        img_h, img_w = img_shape

        # flake8: noqa
        def work_func(_i: int) -> np.ndarray:
            if max_shift_x != 0:
                shift_x = shift_rng.integers(-max_shift_x, max_shift_x)
            else:
                shift_x = 0

            if max_shift_y != 0:
                shift_y = shift_rng.integers(-max_shift_y, max_shift_y)
            else:
                shift_y = 0

            # Total offset: offset to center + random shift
            offset_x = center_x + shift_x
            offset_y = center_y + shift_y

            mask = resize(grid[_i], up_size, order=1)
            mask = mask[offset_y : offset_y + img_h, offset_x : offset_x + img_w]

            return mask

        if threads is None or threads < 1:
            for i in range(num_masks):
                masks[i, ...] = work_func(i)
        else:
            for i, m in enumerate(parallel_map(work_func, range(num_masks), cores=threads, use_multiprocessing=False)):
                masks[i, ...] = m

        return masks

    def get_config(self) -> Dict[str, Any]:
        return {
            "n": self.n,
            "s": self.s,
            "p1": self.p1,
            "seed": self.seed,
            "threads": self.threads,
        }
