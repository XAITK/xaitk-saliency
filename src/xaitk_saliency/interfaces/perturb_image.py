import abc

import numpy as np
from smqtk_core import Plugfigurable


class PerturbImage(Plugfigurable):
    """
    Interface abstracting the behavior of taking a reference image and
    generating some number perturbations in the form of mask matrices
    indicating where perturbations should occur and to what amount.

    Implementations should impart no side effects upon the input image.
    """

    @abc.abstractmethod
    def perturb(self, ref_image: np.ndarray) -> np.ndarray:
        """
        Transform an input reference image into a number of mask matrices
        indicating the perturbed regions.

        Output mask matrix should be three-dimensional with the format
        [nMasks x Height x Width], sharing the same height and width to the
        input reference image.
        The implementing algorithm may determine the quantity of output masks
        per input image.
        These masks should indicate the regions in the corresponding perturbed
        image that have been modified.
        Values should be in the [0, 1] range, where a value closer to 1.0
        indicates areas of the image that are unperturbed.
        Note that output mask matrices may be of a floating-point type
        to allow for fractional perturbation.

        :param ref_image:
            Reference image to generate perturbations from.
        :return: Mask matrix with shape [nMasks x Height x Width].
        """

    def __call__(self, ref_image: np.ndarray) -> np.ndarray:
        """Alias for :meth:`.PerturbImage.perturb`."""
        return self.perturb(ref_image)
