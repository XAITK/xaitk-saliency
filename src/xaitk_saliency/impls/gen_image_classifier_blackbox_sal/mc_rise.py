"""Implementation of MC-RISE saliency stack"""

import itertools
from collections.abc import Generator, Iterable, Sequence
from typing import Any

import numpy as np
from smqtk_classifier import ClassifyImage
from smqtk_descriptors.utils import parallel_map
from typing_extensions import override

from xaitk_saliency.impls.gen_classifier_conf_sal.mc_rise_scoring import MCRISEScoring
from xaitk_saliency.impls.perturb_image.mc_rise import MCRISEGrid
from xaitk_saliency.interfaces.gen_image_classifier_blackbox_sal import GenerateImageClassifierBlackboxSaliency


class MCRISEStack(GenerateImageClassifierBlackboxSaliency):
    """
    Encapsulation of the perturbation-occlusion method using specifically the
    MC-RISE implementations of the component algorithms.

    This more specifically encapsulates the MC-RISE method as presented
    in their paper and code. See references in the :class:`MCRISEGrid`
    and :class:`MCRISEScoring` documentation.

    This implementation shares the `p1` probability and 'k' number colors
    with the internal `MCRISEScoring` instance use, to make use of the
    debiasing described in the MC-RISE paper. Debiasing is always on.
    """

    def __init__(
        self,
        n: int,
        s: int,
        p1: float,
        fill_colors: Sequence[Sequence[int]],
        seed: int = None,
        threads: int = 0,
    ) -> None:
        """
        :param n: int
            Number of random masks used in the algorithm. E.g. 1000.
        :param s: int
            Spatial resolution of the small masking grid. E.g. 8.
            Assumes square grid.
        :param p1: float
            Probability of the grid cell being set to 1 (otherwise 0).
            This should be a float value in the [0, 1] range. E.g. 0.5.
        :param fill_colors: Sequence[Sequence[int]]
            The fill colors to be used when generating masks.
        :param seed: Optional[int]
            A seed to pass into the constructed random number generator to allow
            for reproducibility
        :param threads: int
            The number of threads to utilize when generating masks.
            If this is <=0 or None, no threading is used and processing
            is performed in-line serially.

        :raises: ValueError
            If no fill colors are provided.
        :raises: ValueError
            If provided fill colors have differing numbers of channels.
        """
        if len(fill_colors) == 0:
            raise ValueError("At least one fill color must be provided")
        for fill_color in fill_colors:
            if len(fill_color) != len(fill_colors[0]):
                raise ValueError("All fill colors must have the same number of channels")

        self._perturber = MCRISEGrid(n=n, s=s, p1=p1, k=len(fill_colors), seed=seed, threads=threads)
        self._generator = MCRISEScoring(k=len(fill_colors), p1=p1)
        self._threads = threads
        self._fill_colors = fill_colors

    @staticmethod
    def _work_func(*, ref_image: np.ndarray, i_: int, m: np.ndarray, f: np.ndarray) -> np.ndarray:
        s: tuple = (...,)
        if ref_image.ndim > 2:
            s = (..., None)  # add channel axis for multiplication

        # Just the [H x W] component.
        img_shape = ref_image.shape[:2]

        m_shape = m.shape
        if m_shape != img_shape:
            raise ValueError(
                f"Input mask (position {i_}) did not the shape of the input image: {m_shape} != {img_shape}",
            )
        img_m = np.empty_like(ref_image)

        np.add((m[s] * ref_image), f, out=img_m, casting="unsafe")

        return img_m

    @staticmethod
    def _occlude_image_streaming(
        *,
        ref_image: np.ndarray,
        masks: Iterable[np.ndarray],
        fill_values: Iterable[np.ndarray],
        threads: int = None,
    ) -> Generator[np.ndarray, None, None]:
        if threads is None or threads < 1:
            for i, (mask, fill) in enumerate(zip(masks, fill_values)):
                yield MCRISEStack._work_func(ref_image=ref_image, i_=i, m=mask, f=fill)
        else:
            yield from parallel_map(
                MCRISEStack._work_func,
                ref_image=ref_image,
                i_=itertools.count(),
                m=masks,
                f=fill_values,
                cores=threads,
                use_multiprocessing=False,
            )

    @override
    def _generate(self, ref_image: np.ndarray, blackbox: ClassifyImage) -> np.ndarray:
        """
        Warning: this implementation returns a different shape than is typically expected by this interface.
        Instead of returning `[nClasses x H x W]`, `[kColors x nClasses x H x W] saliency maps will be returned.

        :param ref_image: np.ndarray
            Reference image over which visual saliency heatmaps will be generated.
        :param blackbox: ClassifyImage
            The black-box classifier handle to perform arbitrary operations on in order to deduce visual saliency.

        :return: np.ndarray
            A number of visual saliency heatmaps equivalent in number to the quantity of class labels output
            by the black-box classifier and configured number of colors.
        """
        k_perturbation_masks = self._perturber(ref_image)

        # Collapse k individual colored masks into n multi-color masks
        perturbation_masks = 1 - np.sum(1 - k_perturbation_masks, axis=0)
        fill_values = np.zeros((*k_perturbation_masks.shape[1:], len(self._fill_colors[0])))
        for mask_idx, k_masks in enumerate(np.swapaxes(k_perturbation_masks, 0, 1)):
            k_masks = 1 - np.repeat(k_masks[..., np.newaxis], len(self._fill_colors[0]), axis=3)
            for fill_idx, fill_value in enumerate(self._fill_colors):
                fill_values[mask_idx] += k_masks[fill_idx] * fill_value
        fill_values = np.clip(fill_values, 0, 255)

        class_list = blackbox.get_labels()
        # Input one thing so assume output of one thing.
        ref_conf_dict = list(blackbox.classify_images([ref_image]))[0]
        ref_conf_vec = np.asarray([ref_conf_dict[la] for la in class_list])
        pert_conf_mat = np.empty((perturbation_masks.shape[0], ref_conf_vec.shape[0]), dtype=ref_conf_vec.dtype)

        pert_conf_it = blackbox.classify_images(
            MCRISEStack._occlude_image_streaming(
                ref_image=ref_image,
                masks=perturbation_masks,
                fill_values=fill_values,
                threads=self._threads,
            ),
        )
        for i, pc in enumerate(pert_conf_it):
            pert_conf_mat[i] = [pc[la] for la in class_list]

        # Compose classification results into a matrix for the generator
        # algorithm.
        return self._generator(
            image_conf=ref_conf_vec,
            perturbed_conf=pert_conf_mat,
            perturbed_masks=k_perturbation_masks,
        )

    @override
    def get_config(self) -> dict[str, Any]:
        # It turns out that our configuration here is nearly equivalent to that given
        # and retrieved from the MCRISEGrid implementation
        c = self._perturber.get_config()
        del c["k"]
        c["fill_colors"] = self._fill_colors
        return c
