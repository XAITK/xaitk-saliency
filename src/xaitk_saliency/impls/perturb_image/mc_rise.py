"""Implementation of MC-RISE perturbation mask generation"""

from typing import Any, Optional
from typing_extensions import override

import numpy as np
from skimage.transform import resize
from smqtk_descriptors.utils import parallel_map

from xaitk_saliency.interfaces.perturb_image import PerturbImage


class MCRISEGrid(PerturbImage):
    """
    Based on Hatakeyama et. al:
    https://openaccess.thecvf.com/content/ACCV2020/papers/Hatakeyama_Visualizing_Color-wise_Saliency_of_Black-Box_Image_Classification_Models_ACCV_2020_paper.pdf
    """

    def __init__(
        self,
        n: int,
        s: int,
        p1: float,
        k: int,
        seed: int = None,
        threads: Optional[int] = 4,
    ) -> None:
        """
        Generate a set of random binary masks

        :param n: int
            Number of random masks used in the algorithm. E.g. 1000.
        :param s: int
            Spatial resolution of the small masking grid. E.g. 8.
            Assumes square grid.
        :param p1: float
            Probability of the grid cell being set to 1 (otherwise 0).
            This should be a float value in the [0, 1] range. E.g. 0.5.
        :param k: int
            Number of colors to use.
        :param seed: Optional[int]
            A seed to pass into the constructed random number generator to allow
            for reproducibility
        :param threads: int
            The number of threads to utilize when generating masks.
            If this is <=0 or None, no threading is used and processing
            is performed in-line serially.

        :raises: ValueError
            If p1 not in [0, 1].
        :raises: ValueError
            If k < 1.
        """
        if p1 < 0 or p1 > 1:
            raise ValueError(f"Input p1 value of {p1} is not within the expected [0,1] range.")

        if k <= 0:
            raise ValueError(f"Input k value of {k} is not within the expected >0 range.")

        self.n = n
        self.s = s
        self.p1 = p1
        self.k = k
        self.seed = seed
        self.threads = threads

        # Generate a set of random grids of small resolution
        rng = np.random.default_rng(seed)
        grid: np.ndarray = rng.random((n, s, s)) < p1

        indiv_color_masks = np.empty((self.k, self.n, self.s, self.s))
        for g_idx, g in enumerate(grid):
            g_shape = g.shape
            # Randomly choose fill colors for each pixel
            fill_mask = rng.integers(0, k, size=g_shape)
            for fill_idx in range(self.k):
                # Keep the masked pixels where the current idx (color) was selected
                indiv_color_mask = np.where(g == False, np.where(fill_mask == fill_idx, False, True), True)
                indiv_color_masks[fill_idx][g_idx] = indiv_color_mask

        self.grid = indiv_color_masks
        self.grid.astype("float32")

    @override
    def perturb(self, ref_image: np.ndarray) -> np.ndarray:
        """
        Warning: this implementation returns a different shape than typically expected by this interface.
        Instead of `[nMasks x Height x Width]`, masks of shape `[kColors x nMasks x Height x Width]`
        are returned.

        :param ref_image: np.ndarray
            Reference image to generate perturbations from.

        :return: np.ndarray
            Mask matrix with shape `[kColors x nMasks x Height x Width]`.
        """
        input_size = np.shape(ref_image)[:2]
        num_colors = self.k
        num_masks = self.n
        grid = self.grid
        s = self.s
        shift_rng = np.random.default_rng(self.seed)
        # Shape format: [H x W], Inherits from `input_size`
        cell_size = np.ceil(np.array(input_size) / s)
        up_size = (s + 1) * cell_size

        masks = np.empty((num_colors, num_masks, *input_size), dtype=grid.dtype)

        # Expanding index accesses for repetition efficiency.
        cell_h, cell_w = cell_size[:2]
        input_h, input_w = input_size[:2]

        # flake8: noqa
        def work_func(i_: int) -> np.ndarray:
            # Random shifts
            y = shift_rng.integers(0, cell_h)
            x = shift_rng.integers(0, cell_w)
            k_masks = np.empty((num_colors, input_h, input_w))
            for k_ in range(num_colors):
                k_masks[k_] = resize(
                    grid[k_, i_, ...],
                    up_size,
                    order=1,
                    mode="reflect",
                    anti_aliasing=False,
                )[
                    y : y + input_h,
                    x : x + input_w,
                ]
            return k_masks

        threads = self.threads
        if threads is None or threads < 1:
            for i in range(num_masks):
                masks[:, i, ...] = work_func(i)
        else:
            for i, m in enumerate(
                parallel_map(
                    work_func,
                    range(num_masks),
                    cores=threads,
                    use_multiprocessing=False,
                ),
            ):
                masks[:, i, ...] = m

        return masks

    @override
    def get_config(self) -> dict[str, Any]:
        return {
            "n": self.n,
            "s": self.s,
            "p1": self.p1,
            "k": self.k,
            "seed": self.seed,
            "threads": self.threads,
        }
