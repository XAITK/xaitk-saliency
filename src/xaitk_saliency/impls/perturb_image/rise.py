"""
RISE MIT License (https://github.com/eclique/RISE/blob/master/LICENSE):

    MIT License

    Copyright (c) 2018 Vitali Petsiuk

    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in all
    copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
    SOFTWARE.
"""

from typing import Any, Dict, Optional

import numpy as np
from skimage.transform import resize
from smqtk_descriptors.utils import parallel_map

from xaitk_saliency.interfaces.perturb_image import PerturbImage


class RISEGrid(PerturbImage):
    """
    Based on Petsiuk et. al: http://bmvc2018.org/contents/papers/1064.pdf

    Implementation is borrowed from the original authors:
    https://github.com/eclique/RISE/blob/master/explanations.py
    """

    def __init__(
        self,
        n: int,
        s: int,
        p1: float,
        seed: Optional[int] = None,
        threads: Optional[int] = 4,
    ):
        """
        Generate a set of random binary masks

        :param n:
            Number of random masks used in the algorithm. E.g. 1000.
        :param s:
            Spatial resolution of the small masking grid. E.g. 8.
            Assumes square grid.
        :param p1:
            Probability of the grid cell being set to 1 (otherwise 0).
            This should be a float value in the [0, 1] range. E.g. 0.5.
        :param seed:
            A seed to pass into the constructed random number generator to allow
            for reproducibility
        :param threads: The number of threads to utilize when generating masks.
            If this is <=0 or None, no threading is used and processing
            is performed in-line serially.
        """
        if p1 < 0 or p1 > 1:
            raise ValueError(f"Input p1 value of {p1} is not within the expected [0,1] range.")

        self.n = n
        self.s = s
        self.p1 = p1
        self.seed = seed
        self.threads = threads

        # Generate a set of random grids of small resolution
        grid: np.ndarray = np.random.default_rng(seed).random((n, s, s)) < p1
        grid = grid.astype("float32")

        self.grid = grid

    def perturb(self, ref_image: np.ndarray) -> np.ndarray:
        input_size = np.shape(ref_image)[:2]
        num_masks = self.n
        grid = self.grid
        s = self.s
        shift_rng = np.random.default_rng(self.seed)
        # Shape format: [H x W], Inherits from `input_size`
        cell_size = np.ceil(np.array(input_size) / s)
        up_size = (s + 1) * cell_size

        masks = np.empty((num_masks, *input_size), dtype=grid.dtype)

        # Expanding index accesses for repetition efficiency.
        cell_h, cell_w = cell_size[:2]
        input_h, input_w = input_size[:2]

        # flake8: noqa
        def work_func(i_: int) -> np.ndarray:
            # Random shifts
            y = shift_rng.integers(0, cell_h)
            x = shift_rng.integers(0, cell_w)
            mask = resize(
                grid[i_],
                up_size,
                order=1,
                mode="reflect",
                anti_aliasing=False,
            )[
                y : y + input_h,
                x : x + input_w,
            ]
            return mask

        threads = self.threads
        if threads is None or threads < 1:
            for i in range(num_masks):
                masks[i, ...] = work_func(i)
        else:
            for i, m in enumerate(
                parallel_map(
                    work_func,
                    range(num_masks),
                    cores=threads,
                    use_multiprocessing=False,
                ),
            ):
                masks[i, ...] = m

        return masks

    def get_config(self) -> Dict[str, Any]:
        return {
            "n": self.n,
            "s": self.s,
            "p1": self.p1,
            "seed": self.seed,
            "threads": self.threads,
        }
