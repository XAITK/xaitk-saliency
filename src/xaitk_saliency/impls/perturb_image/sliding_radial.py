from typing import Any, Optional

import numpy as np
from scipy.ndimage.filters import gaussian_filter
from skimage.draw import ellipse

from xaitk_saliency.interfaces.perturb_image import PerturbImage


class SlidingRadial(PerturbImage):
    """
    Produce perturbation matrices generated by sliding a radial occlusion area
    with configured radius over the area of an image. When the two radius values
    are the same, circular masks are generated; otherwise, elliptical masks are
    generated. Passing sigma values will apply a Gaussian filter to the mask,
    blurring it. This results in a smooth transition from full occlusion in the
    center of the radial to no occlusion at the edge.

    Due to the geometry of sliding radials, if the stride given does not evenly
    divide the radial size along the applicable axis, then the result plane of
    values when summing the generated masks will not be even.

    Related, if the stride is set to be larger than the radial diameter, the
    resulting plane of summed values will also not be even, as there be
    increasingly long valleys of unperturbed space between masked regions.

    The generated masks are boolean if no blurring is used, otherwise the masks
    will be of floating-point type in the [0, 1] range.

    :param radius: The radius of the occlusion area in pixels as a tuple with
        format `(radius_y, radius_x)`.
    :param stride: The striding step in pixels for the center of the radial as
        a tuple with format `(height_step, width_step)`.
    :param sigma: The sigma values for the Gaussian filter applied to masks in
        pixels as a tuple with format `(sigma_y, sigma_x)`.
    """

    def __init__(
        self,
        radius: tuple[float, float] = (50, 50),
        stride: tuple[int, int] = (20, 20),
        sigma: Optional[tuple[float, float]] = None,
    ) -> None:
        self.radius = (radius[0], radius[1])
        self.stride = (int(stride[0]), int(stride[1]))
        self.sigma = (sigma[0], sigma[1]) if sigma else None

    def perturb(self, ref_image: np.ndarray) -> np.ndarray:
        stride_h, stride_w = self.stride
        img_h, img_w = ref_image.shape[:2]
        center_xs = np.arange(0, img_w, stride_w)
        center_ys = np.arange(0, img_h, stride_h)

        num_masks = len(center_xs) * len(center_ys)
        masks = np.zeros((num_masks, img_h, img_w), dtype="float32" if self.sigma else "bool")
        center_xs_m = np.tile(center_xs, len(center_ys))
        center_ys_m = np.repeat(center_ys, len(center_xs))

        for i, (center_x, center_y) in enumerate(zip(center_xs_m, center_ys_m)):
            mask = masks[i]
            coords = ellipse(center_y, center_x, *self.radius, shape=mask.shape)
            mask[coords] = 1

            if self.sigma:
                mask[:] = gaussian_filter(mask, sigma=self.sigma)
                mask[:] = mask / mask.max()

            mask[:] = 1 - mask if self.sigma else ~mask

        return masks

    @classmethod
    def get_default_config(cls) -> dict[str, Any]:
        # Minor override to curry tuple defaults into lists, which are the
        # JSON-parsed types. This is to allow successful equality between
        # default, get_config() and JSON-parsed outputs.
        cfg = super().get_default_config()
        cfg["radius"] = list(cfg["radius"])
        cfg["stride"] = list(cfg["stride"])
        return cfg

    def get_config(self) -> dict[str, Any]:
        return {
            "radius": list(self.radius),
            "stride": list(self.stride),
            "sigma": list(self.sigma) if self.sigma else None,
        }
