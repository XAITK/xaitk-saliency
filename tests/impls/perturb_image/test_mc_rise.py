import numpy as np
import pytest
from smqtk_core.configuration import configuration_test_helper
from syrupy.assertion import SnapshotAssertion

from tests.test_utils import CustomFloatSnapshotExtension
from xaitk_saliency.impls.perturb_image.mc_rise import MCRISEGrid


@pytest.fixture
def snapshot_custom(snapshot: SnapshotAssertion) -> SnapshotAssertion:
    return snapshot.use_extension(CustomFloatSnapshotExtension)


class TestMCRISEPerturbation:
    def test_init_valued(self) -> None:
        """Test that constructor values pass."""
        ex_n = 1000
        ex_s = 8
        ex_p1 = 0.5
        k = 2
        impl = MCRISEGrid(n=ex_n, s=ex_s, p1=ex_p1, k=k)
        assert impl.n == ex_n
        assert impl.s == ex_s
        assert np.allclose(impl.p1, ex_p1)
        assert impl.k == k

    def test_init_outofrange_p1(self) -> None:
        """Test catching an out of range p1 value."""
        with pytest.raises(ValueError, match=r"Input p1 value of -0\.3 is not within the expected \[0,1\] range\."):
            MCRISEGrid(10, 8, p1=-0.3, k=1)

        with pytest.raises(ValueError, match=r"Input p1 value of 5 is not within the expected \[0,1\] range\."):
            MCRISEGrid(10, 8, p1=5, k=1)

        with pytest.raises(ValueError, match=r"Input k value of -1 is not within the expected >0 range\."):
            MCRISEGrid(10, 8, p1=0.2, k=-1)

    def test_standard_config(self) -> None:
        ex_n = 1000
        ex_s = 8
        ex_p1 = 0.5
        k = 1
        impl = MCRISEGrid(n=ex_n, s=ex_s, p1=ex_p1, k=k)
        for inst in configuration_test_helper(impl):
            assert inst.n == ex_n
            assert inst.s == ex_s
            assert np.allclose(inst.p1, ex_p1)
            assert inst.k == k

    def test_if_random(self) -> None:
        """Test that the perturbations are randomized"""
        impl1 = MCRISEGrid(n=1000, s=8, p1=0.5, k=2)
        impl2 = MCRISEGrid(n=1000, s=8, p1=0.5, k=2)
        assert not np.array_equal(impl1.grid, impl2.grid)

    def test_seed(self) -> None:
        """Test that passing a seed generates equivalent masks"""
        impl1 = MCRISEGrid(n=1000, s=8, p1=0.5, k=2, seed=42)
        impl2 = MCRISEGrid(n=1000, s=8, p1=0.5, k=2, seed=42)
        assert np.array_equal(impl1.grid, impl2.grid)

    def test_perturb_1channel(self, snapshot_custom: SnapshotAssertion) -> None:
        """
        Test basic perturbation on a known image with even windowing + stride.
        Input image mode should not impact the masks output.
        """
        # Image is slightly wide
        white_image = np.full((4, 6), fill_value=255, dtype=np.uint8)

        # Setting threads=0 for serialized processing for deterministic
        # results.
        impl = MCRISEGrid(n=2, s=2, p1=0.5, k=2, seed=42, threads=0)
        actual_masks = impl.perturb(white_image)

        snapshot_custom.assert_match(actual_masks)

    def test_call_idempotency(self) -> None:
        """
        Test that, at least when seeded and single-threaded, perturbation
        generation is idempotent.
        """
        # Image is slightly wide
        white_image = np.full((4, 6), fill_value=255, dtype=np.uint8)
        # Setting threads=0 for serialized processing for deterministic
        # results. When greater than 1 idempotency cannot be guaranteed due to
        # thread interleaving.
        # Also of course seeding otherwise random will do its random things.
        impl = MCRISEGrid(n=2, s=2, p1=0.5, k=3, seed=42, threads=0)

        actual_masks1 = impl.perturb(white_image)
        actual_masks2 = impl.perturb(white_image)

        assert np.allclose(
            actual_masks1,
            actual_masks2,
        )

    def test_perturb_3channel(self, snapshot_custom: SnapshotAssertion) -> None:
        """
        Test basic perturbation on a known image with even windowing + stride.
        Input image mode should not impact the masks output.
        """
        # Image is slightly wide
        white_image = np.full((4, 6, 3), fill_value=255, dtype=np.uint8)

        # Setting threads=0 for serialized processing for deterministic
        # results.
        impl = MCRISEGrid(n=2, s=2, p1=0.5, k=2, seed=42, threads=0)
        actual_masks = impl.perturb(white_image)

        snapshot_custom.assert_match(actual_masks)

    def test_multiple_image_sizes(self) -> None:
        """
        Test that once we initialize a RISEPerturbation we can call it on
        images of varying sizes
        """
        impl = MCRISEGrid(n=2, s=2, p1=0.5, k=2, seed=42)
        white_image_small = np.full((4, 6), fill_value=255, dtype=np.uint8)
        white_image_large = np.full((41, 26), fill_value=255, dtype=np.uint8)
        masks_small = impl.perturb(white_image_small)
        assert len(masks_small) == 2
        assert masks_small.shape[2:] == white_image_small.shape

        masks_large = impl.perturb(white_image_large)
        assert len(masks_large) == 2
        assert masks_large.shape[2:] == white_image_large.shape
