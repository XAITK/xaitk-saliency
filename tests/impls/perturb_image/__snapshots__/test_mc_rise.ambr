# serializer version: 1
# name: TestMCRISEPerturbation.test_perturb_1channel
  array([[[[0.98148148, 0.90740741, 0.83333333, 0.75925926, 0.68518519,
            0.72222222],
           [1.        , 1.        , 1.        , 1.        , 1.        ,
            1.        ],
           [0.98148148, 0.90740741, 0.83333333, 0.75925926, 0.68518519,
            0.72222222],
           [0.96296296, 0.81481481, 0.66666667, 0.51851852, 0.37037037,
            0.44444444]],

          [[0.83333333, 0.94444444, 0.72222222, 0.5       , 0.27777778,
            0.05555556],
           [0.61111111, 0.64814815, 0.57407407, 0.5       , 0.42592593,
            0.35185185],
           [0.38888889, 0.35185185, 0.42592593, 0.5       , 0.57407407,
            0.64814815],
           [0.16666667, 0.05555556, 0.27777778, 0.5       , 0.72222222,
            0.94444444]]],


         [[[0.05555556, 0.27777778, 0.5       , 0.72222222, 0.94444444,
            0.83333333],
           [0.05555556, 0.27777778, 0.5       , 0.72222222, 0.94444444,
            0.83333333],
           [0.05555556, 0.27777778, 0.5       , 0.72222222, 0.94444444,
            0.83333333],
           [0.05555556, 0.27777778, 0.5       , 0.72222222, 0.94444444,
            0.83333333]],

          [[1.        , 1.        , 1.        , 1.        , 1.        ,
            1.        ],
           [0.94444444, 0.98148148, 0.90740741, 0.83333333, 0.75925926,
            0.68518519],
           [0.88888889, 0.96296296, 0.81481481, 0.66666667, 0.51851852,
            0.37037037],
           [0.83333333, 0.94444444, 0.72222222, 0.5       , 0.27777778,
            0.05555556]]]])
# ---
# name: TestMCRISEPerturbation.test_perturb_3channel
  array([[[[0.98148148, 0.90740741, 0.83333333, 0.75925926, 0.68518519,
            0.72222222],
           [1.        , 1.        , 1.        , 1.        , 1.        ,
            1.        ],
           [0.98148148, 0.90740741, 0.83333333, 0.75925926, 0.68518519,
            0.72222222],
           [0.96296296, 0.81481481, 0.66666667, 0.51851852, 0.37037037,
            0.44444444]],

          [[0.83333333, 0.94444444, 0.72222222, 0.5       , 0.27777778,
            0.05555556],
           [0.61111111, 0.64814815, 0.57407407, 0.5       , 0.42592593,
            0.35185185],
           [0.38888889, 0.35185185, 0.42592593, 0.5       , 0.57407407,
            0.64814815],
           [0.16666667, 0.05555556, 0.27777778, 0.5       , 0.72222222,
            0.94444444]]],


         [[[0.05555556, 0.27777778, 0.5       , 0.72222222, 0.94444444,
            0.83333333],
           [0.05555556, 0.27777778, 0.5       , 0.72222222, 0.94444444,
            0.83333333],
           [0.05555556, 0.27777778, 0.5       , 0.72222222, 0.94444444,
            0.83333333],
           [0.05555556, 0.27777778, 0.5       , 0.72222222, 0.94444444,
            0.83333333]],

          [[1.        , 1.        , 1.        , 1.        , 1.        ,
            1.        ],
           [0.94444444, 0.98148148, 0.90740741, 0.83333333, 0.75925926,
            0.68518519],
           [0.88888889, 0.96296296, 0.81481481, 0.66666667, 0.51851852,
            0.37037037],
           [0.83333333, 0.94444444, 0.72222222, 0.5       , 0.27777778,
            0.05555556]]]])
# ---
